README.md
Notrequests
===========

A Python wrapper for the built-in urllib module. The API is compatible with [the excellent Requests library][requests], but omitting features such as sessions and support for keep-alive, all in a single module (but requires the [six compatibility library][six]).

Notrequests is intended for doing HTTP requests on [Google App Engine][gae] where Requests has some disadvantages. It works on Python 2.7 and Python 3.4 and later.

The project is [hosted on GitHub][notrequests].


Installation
------------

From PyPI:

    $ pip install notrequests

Or download and run setup as normal:

    $ curl -L -o notrequests.zip https://github.com/davidwtbuxton/notrequests/archive/master.zip
    $ unzip notrequests.zip
    $ cd notrequests-master
    $ python setup.py install


Usage
-----


### Basic usage

Notrequests is compatible with the Requests API (or it tries to be).

    >>> import notrequests
    >>>
    >>> response = notrequests.get('http://httpbin.org/get')
    >>> response.status_code == notrequests.codes.ok
    True

But it doesn't do everything that Requests does. There's no session support, no keep-alive support and it reads the entire response into memory.

The response body is available as a byte string or as unicode.

    >>> response = notrequests.get('http://httpbin.org/encoding/utf8')
    >>> response.headers['content-type']
    'text/html; charset=utf-8'
    >>> type(response.content)
    <type 'str'>
    >>> type(response.text)
    <type 'unicode'>

Decoding to unicode relies on the server having sent a valid content-type header. This is different to Requests because Requests has smarts to sniff the encoding should the response not include a content-type header.

Notrequests uses urllib2 but behaves more like Requests. So it won't throw an exception on 4xx and 5xx responses.

    >>> response = notrequests.get('http://httpbin.org/status/404')
    >>> response.status_code == notrequests.codes.not_found
    True

You can also test for failure, or raise an exception.

    >>> response = notrequests.get('http://httpbin.org/status/200')
    >>> response.ok
    True
    >>> response.raise_for_status()
    >>> response = notrequests.get('http://httpbin.org/status/404')
    >>> response.ok
    False
    >>> response.raise_for_status()
    Traceback (most recent call last):
      ...
    notrequests.HTTPError: Error 404 for http://httpbin.org/status/404


### Redirects

If you want to prevent Notrequests following a redirect response, you can use the `allow_redirects` keyword:

    >>> url = 'http://httpbin.org/redirect/1'
    >>> response = notrequests.get(url)
    >>> response.status_code
    200
    >>> response = notrequests.get(url, allow_redirects=False)
    >>> response.status_code
    302

On Google App Engine, the `X-Appengine-Inbound-Appid` header will only be set if [the sending application doesn't allow redirects!][appidentity]


### Authentication

You can do basic auth just like Requests (but not other authentication types):

    >>> url = 'http://httpbin.org/basic-auth/alice/secret'
    >>> response = notrequests.get(url)
    >>> response.status_code
    401
    >>> response = notrequests.get(url, auth=('alice', 'secret'))
    >>> response.status_code
    200


### JSON

And send and decode JSON:

    >>> import pprint
    >>> response = notrequests.put('http://httpbin.org/put', json={'foo': ['bar', 'baz']})
    >>> data = response.json()
    >>> pprint.pprint(data)
    {u'args': {},
     u'data': u'{"foo": ["bar", "baz"]}',
     u'files': {},
     u'form': {},
     u'headers': {u'Accept-Encoding': u'identity',
                  u'Content-Length': u'23',
                  u'Content-Type': u'application/json',
                  u'Host': u'httpbin.org',
                  u'User-Agent': u'notrequests/0.1'},
     u'json': {u'foo': [u'bar', u'baz']},
     u'origin': u'10.10.10.1',
     u'url': u'http://httpbin.org/put'}


### Accessing link headers

If the server sent 'Link' headers in the response (often used by APIs to give links to the next page of results) then you can get the parsed links straight from the response object:

    >>> response.headers['Link']
    '<https://example.com/?page=2>; rel="next"'
    >>> response.links['next']['url']
    'https://example.com/?page=2'


### Uploading files

There's also support for uploading files:

    >>> import io
    >>> fileobj = io.BytesIO('foo bar baz')
    >>> response = notrequests.post('http://httpbin.org/post', files={'upload': fileobj})
    >>> response.json()['files']
    {u'upload': 'foo bar baz'}

As with Requests, the keys in the files dict are the form field input names and
the values in the files dict can be a 2-tuple of file name with file object or
byte string:

    >>> files = {'upload': ('my-file.txt', b'Foo\nbar\nbaz.')}
    >>> response = notrequests.post('http://httpbin.org/post', files=files)
    >>> print response.request.data
    --10.10.10.1.503.2717.1443987498.810.2
    Content-Disposition: file; name="upload"; filename="my-file.txt"
    Content-Type: text/plain

    Foo
    bar
    baz.
    --10.10.10.1.503.2717.1443987498.810.2--


### Disabling SSL certificate checking

Use the `verify` keyword to disable SSL certificate checks. The default is `verify=True`, so Notrequests will raise `ssl.CertificateError` if the certificate does not match the server's hostname.

    >>> response = notrequests.get('https://swupdl.adobe.com', verify=False)

Notrequests does not support specifying alternate CA bundles.


API compatibility
-----------------

These are some features of [the Requests API][api] that Notrequests has _not_ implemented. It isn't a complete list, and it would be nice to have better support.

- Sessions
- Response.history
- Streaming uploads / downloads and iterating over data
- Alternate names for status codes
- Proxies

If there is a missing feature that you want to use on App Engine, please [open an issue on GitHub][issues].


Tests
-----

Run the tests with [tox][tox].

By default the tests make requests to http://httpbin.org, but you can run a local instance which will speed things up.

    $ pip install httpbin gunicorn
    $ gunicorn --bind 127.0.0.1:8888 httpbin:app &
    $ export NOTREQUESTS_TEST_URL="http://127.0.0.1:8888"
    $ tox


Why not use Requests?
---------------------

Google App Engine patches httplib in the standard library to use its urlfetch service, and restricts [the sockets API][sockets] to applications with billing enabled. Requests does not use httplib and uses sockets.

If you want to use [the app identity service to authenticate connections between App Engine applications][appidentity] you have to use [the urlfetch service][urlfetch], you cannot use Requests. Notrequests works on App Engine, using the urlfetch service, because it uses the standard library.


[notrequests]: https://github.com/davidwtbuxton/notrequests
[requests]: http://www.python-requests.org/
[gae]: https://cloud.google.com/appengine/
[tox]: http://codespeak.net/tox/
[appidentity]: https://cloud.google.com/appengine/docs/python/appidentity/#asserting_identity_to_other_app_engine_apps
[sockets]: https://cloud.google.com/appengine/docs/python/sockets/
[api]: http://requests.readthedocs.org/en/latest/api/
[six]: https://pythonhosted.org/six/
[urlfetch]: https://cloud.google.com/appengine/docs/python/outbound-requests
[issues]: https://github.com/davidwtbuxton/notrequests/issues


NEWS.md
0.7 - 6 July 2016
-----------------

- Added support for Python 3.4 and 3.5, requiring the six compatibility library.

0.6 - 29 June 2016
------------------

- Implemented `response.ok` and `response.raise_for_status()` (#5).

0.5 - 16 April 2016
-------------------

- Fixed decoding of cookies which set a domain or path or expiry date (#4).
- Added support for specifying a timeout (#3).


tests/test_notrequests.py
#!/usr/bin/env python
import io
import json
import os
import socket
import ssl
import tempfile
import unittest
import warnings

import six
from six.moves import urllib

import notrequests as nr


def _url(path):
    # See README on how to use a local httpbin instance for testing.
    base_url = os.environ.get('NOTREQUESTS_TEST_URL', 'http://httpbin.org/')
    return urllib.parse.urljoin(base_url, path)


class PackageAPITestCase(unittest.TestCase):
    def test_api(self):
        nr.get
        nr.patch
        nr.post
        nr.put
        nr.delete
        nr.head
        nr.codes
        nr.HTTPError


class GetTestCase(unittest.TestCase):
    def test_get(self):
        url = _url('/')
        response = nr.get(url)

        self.assertEqual(response.status_code, 200)

    def test_auth_get_allowed(self):
        url = _url('/basic-auth/foo/bar')
        response = nr.get(url, auth=('foo', 'bar'))

        self.assertEqual(response.status_code, 200)

    def test_auth_get_denied(self):
        url = _url('/basic-auth/foo/bar')
        response = nr.get(url, auth=('baz', 'qux'))

        self.assertEqual(response.status_code, 401)

    def test_setting_headers_on_get(self):
        url = _url('/headers')
        headers = {'X-Foo': 'bar'}
        response = nr.get(url, headers=headers)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertIn('headers', data)
        self.assertIn('X-Foo', data['headers'])

    def test_headers_case_normalized(self):
        url = _url('/headers')
        headers = {'X-FOO-bar-Baz': 'BAR'}
        response = nr.get(url, headers=headers)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertIn('headers', data)
        self.assertIn('X-Foo-Bar-Baz', data['headers'])

    def test_get_500_response(self):
        url = _url('/status/500')
        response = nr.get(url)

        self.assertEqual(response.status_code, 500)

    def test_sending_cookies(self):
        url = _url('/cookies')
        cookies = {'foo': 'bar'}
        response = nr.get(url, cookies=cookies)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data, {'cookies': {'foo': 'bar'}})

    def test_receiving_cookies(self):
        url = _url('/cookies/set?foo=bar')
        response = nr.get(url, allow_redirects=False)

        self.assertEqual(response.status_code, 302)
        self.assertEqual(response.cookies, {'foo': 'bar'})

    def test_params_dict_sets_query_string(self):
        url = _url('/get')
        params = {'foo': 'bar'}
        response = nr.get(url, params=params)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['args'], params)

    def test_params_merged_with_url_query_string(self):
        url = _url('/get?foo=bar')
        params = {'foo': 'baz'}
        response = nr.get(url, params=params)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['args'], {'foo': ['bar', 'baz']})

    def test_params_tuple_sets_query_string(self):
        url = _url('/get')
        params = (('foo', 'bar'), ('baz', 'qux'),)
        response = nr.get(url, params=params)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['args'], {'foo': 'bar', 'baz': 'qux'})

    def test_verify_ssl_is_true(self):
        url = 'https://swupdl.adobe.com'

        with self.assertRaises(ssl.CertificateError):
            response = nr.get(url, verify=True)

    def test_verify_ssl_is_false(self):
        url = 'https://swupdl.adobe.com'
        response = nr.get(url, verify=False)

        # The important thing is it didn't throw an exception.
        self.assertEqual(response.status_code, 404)

    def test_timeout_keyword(self):
        url = _url('/')
        response = nr.get(url, timeout=60)

    def test_timeout_raises_error(self):
        # On App Engine I think it raises DeadlineExceededError.
        url = _url('/delay/2')

        with self.assertRaises(socket.timeout):
            response = nr.get(url, timeout=1)

    def test_timeout_does_not_raise_error(self):
        url = _url('/delay/1')
        response = nr.get(url, timeout=2)

        self.assertEqual(response.status_code, 200)


class PatchTestCase(unittest.TestCase):
    def test_patch(self):
        url = _url('/patch')
        response = nr.patch(url)

        self.assertEqual(response.status_code, 200)


class PostTestCase(unittest.TestCase):
    def test_sending_data_dict_is_form_encoded(self):
        url = _url('/post')
        request_data = {
            'foo': 'bar baz',
        }
        response = nr.post(url, data=request_data)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['headers']['Content-Type'], 'application/x-www-form-urlencoded')
        self.assertEqual(data['form'], request_data)

    def test_sending_data_as_bytes(self):
        url = _url('/post')
        request_data = json.dumps({'foo': 'bar'}).encode('utf-8')
        headers={'Content-Type': 'application/json'}
        response = nr.post(url, data=request_data, headers=headers)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['headers']['Content-Type'], 'application/json')
        self.assertEqual(data['data'], '{"foo": "bar"}')

    def test_json_keyword(self):
        url = _url('/post')
        json_data = {'foo': 'bar'}
        response = nr.post(url, json=json_data)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['data'], '{"foo": "bar"}')
        self.assertEqual(data['headers']['Content-Type'], 'application/json')

    def test_submit_file_with_file_object(self):
        url = _url('/post')
        files = {'file': io.BytesIO(b'binarydata')}
        response = nr.post(url, files=files)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['files'], {'file': 'binarydata'})

    def test_submit_file_with_name_and_file_object(self):
        url = _url('/post')
        files = {'file': ('foo.txt', io.BytesIO(b'binarydata'))}
        response = nr.post(url, files=files)

        self.assertEqual(response.status_code, 200)

        data = response.json()
        self.assertEqual(data['files'], {'file': 'binarydata'})

    def test_submit_file_with_name_and_byte_string(self):
        url = _url('/post')
        files = {'file': ('foo.txt', b'binarydata')}
        response = nr.post(url, files=files)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['files'], {'file': 'binarydata'})

    def test_submit_file_with_file_from_disk(self):
        # This exercises the code which guesses the uploaded filename from
        # the open file handle.

        url = _url('/post')
        _, path = tempfile.mkstemp()

        with open(path, 'wb') as fh:
            fh.write(b'binarydata')

        try:
            with open(path, 'rb') as fh:
                files = {'file': fh}
                response = nr.post(url, files=files)

            self.assertEqual(response.status_code, 200)

            data = response.json()

            self.assertEqual(data['files'], {'file': 'binarydata'})
        finally:
            os.unlink(path)

    def test_submit_file_and_form_data(self):
        url = _url('/post')
        files = {'file': io.BytesIO(b'binarydata')}
        request_data = {'foo': b'bar baz'}
        response = nr.post(url, files=files, data=request_data)

        self.assertEqual(response.status_code, 200)

        data = response.json()

        self.assertEqual(data['files'], {'file': 'binarydata'})
        self.assertEqual(data['form'], {'foo': 'bar baz'})


class PutTestCase(unittest.TestCase):
    def test_put(self):
        url = _url('/put')
        response = nr.put(url)

        self.assertEqual(response.status_code, 200)


class DeleteTestCase(unittest.TestCase):
    def test_delete(self):
        url = _url('/delete')
        response = nr.delete(url)

        self.assertEqual(response.status_code, 200)


class HeadTestCase(unittest.TestCase):
    def test_head(self):
        url = _url('/get')
        response = nr.head(url)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.content, b'')


class ResponseTestCase(unittest.TestCase):
    def test_decode_json(self):
        url = _url('/user-agent')
        response = nr.get(url)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), {'user-agent': nr._user_agent})

    def test_status_code(self):
        url = _url('/user-agent')
        response = nr.get(url)

        self.assertEqual(response.status_code, 200)

    def test_content(self):
        url = _url('/user-agent')
        response = nr.get(url)
        expected = ('{\n  "user-agent": "%s"\n}\n' % nr._user_agent).encode('latin-1')

        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.content, expected)

    def test_decoding_text_with_encoding_header(self):
        url = _url('/encoding/utf8')
        response = nr.get(url)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.headers['Content-Type'], 'text/html; charset=utf-8')
        self.assertIsInstance(response.text, six.text_type)
        self.assertEqual(response.text[:21], u'<h1>Unicode Demo</h1>')

    def test_links_property_with_valid_link_header(self):
        url = _url('/response-headers')
        link = '<https://example.com/?page=2>; rel="next", <https://example.com/?page=34>; rel="last"'
        response = nr.get(url, params={'Link': link})

        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.headers['Link'], link)
        self.assertEqual(
            response.links,
            {
                'last': {'rel': 'last', 'url': 'https://example.com/?page=34'},
                'next': {'rel': 'next', 'url': 'https://example.com/?page=2'},
            },
        )

    def test_links_property_with_no_link_header(self):
        url = _url('/response-headers')
        response = nr.get(url, params={})

        self.assertEqual(response.status_code, 200)
        self.assertNotIn('Link', response.headers)
        self.assertEqual(response.links, {})

    def test_cookies_are_decoded_correctly(self):
        # Previously we caused cookielib errors trying to parse complex cookies.
        # http://www.adobe.com/support/downloads/product.jsp?product=1&platform=Windows

        url = _url('/response-headers')
        # Domain can changed depending on NOTREQUESTS_TEST_URL being set.
        domain = urllib.parse.urlparse(url).netloc.split(':')[0]

        # httpbin's /cookies/set path doesn't allow us to specify domain, so
        # set the full cookie header value manually.
        cookie_values = [
            'JSESSIONID=ABC123; Path=/',
            'dtCookie=DEF456; Path=/; Domain={}',
            'AWID=GHI789; Version=1; Comment="hello"; Domain={}; Max-Age=315360000; Path=/',
        ]

        with warnings.catch_warnings(record=True) as raised_warnings:
            warnings.simplefilter('always')

            params = [('Set-Cookie', value.format(domain)) for value in cookie_values]
            response = nr.get(url, params=params, allow_redirects=False)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(raised_warnings, [])
        self.assertEqual(
            response.cookies,
            {
                'JSESSIONID': 'ABC123',
                'dtCookie': 'DEF456',
                'AWID': 'GHI789',
            },
        )

    def test_raise_for_status_raises_error_for_404(self):
        url = _url('/status/404')
        response = nr.get(url)

        with self.assertRaisesRegexp(nr.HTTPError, r'Error 404 for %s' % url):
            response.raise_for_status()

    def test_raise_for_status_raises_error_for_500(self):
        url = _url('/status/500')
        response = nr.get(url)

        with self.assertRaisesRegexp(nr.HTTPError, r'Error 500 for %s' % url):
            response.raise_for_status()

    def test_raise_for_status_no_error(self):
        url = _url('/status/200')
        response = nr.get(url)

        self.assertIsNone(response.raise_for_status())

    def test_response_ok_true(self):
        url = _url('/status/200')
        response = nr.get(url)

        self.assertTrue(response.ok)

    def test_response_ok_false(self):
        url = _url('/status/404')
        response = nr.get(url)

        self.assertFalse(response.ok)


class CodesTestCase(unittest.TestCase):
    def test_access_status_codes_as_properties(self):
        self.assertEqual(nr.codes.ok, 200)


class DetectEncodingTestCase(unittest.TestCase):
    def test_detect_utf8(self):
        value = json.dumps({'foo': 'bar'}).encode('utf-8')
        result = nr.detect_encoding(value)

        self.assertEqual(result, 'utf-8')

    def test_detect_utf16le(self):
        value = json.dumps({'foo': 'bar'}).encode('utf-16-le')
        result = nr.detect_encoding(value)

        self.assertEqual(result, 'utf-16-le')

    def test_detect_utf16be(self):
        value = json.dumps({'foo': 'bar'}).encode('utf-16-be')
        result = nr.detect_encoding(value)

        self.assertEqual(result, 'utf-16-be')

    def test_detect_utf32le(self):
        value = json.dumps({'foo': 'bar'}).encode('utf-32-le')
        result = nr.detect_encoding(value)

        self.assertEqual(result, 'utf-32-le')

    def test_detect_utf32be(self):
        value = json.dumps({'foo': 'bar'}).encode('utf-32-be')
        result = nr.detect_encoding(value)

        self.assertEqual(result, 'utf-32-be')


if __name__ == '__main__':
    unittest.main()


notrequests.py
import base64
import functools
import json as simplejson
import mimetypes
import os
import random
import re
import ssl

import six
from six.moves import http_cookiejar
from six.moves import urllib


__version__ = '0.7'
_user_agent = 'notrequests/' + __version__
LATIN1 = 'latin-1'
JSON_TYPE = 'application/json'
BINARY_TYPE = 'application/octet-stream'

_codes = {
    # Informational.
    100: 'continue',
    101: 'switching_protocols',
    102: 'processing',
    200: 'ok',
    201: 'created',
    202: 'accepted',
    203: 'non_authoritative_information',
    204: 'no_content',
    205: 'reset_content',
    206: 'partial_content',

    # Redirection.
    300: 'multiple_choices',
    301: 'moved_permanently',
    302: 'found',
    303: 'see_other',
    304: 'not_modified',
    305: 'use_proxy',
    306: 'switch_proxy',
    307: 'temporary_redirect',

    # Client error.
    400: 'bad_request',
    401: 'unauthorized',
    402: 'payment_required',
    403: 'forbidden',
    404: 'not_found',
    405: 'method_not_allowed',
    406: 'not_acceptable',
    407: 'proxy_authentication_required',
    408: 'request_timeout',
    409: 'conflict',
    410: 'gone',
    411: 'length_required',
    412: 'precondition_failed',
    413: 'request_entity_too_large',
    414: 'request_uri_too_long',
    415: 'unsupported_media_type',
    416: 'requested_range_not_satisfiable',
    417: 'expectation_failed',

    # Server error.
    500: 'internal_server_error',
    501: 'not_implemented',
    502: 'bad_gateway',
    503: 'service_unavailable',
    504: 'gateway_timeout',
    505: 'http_version_not_supported',
}


class PropertyDict(dict):
    def __getattr__(self, name):
        return self.get(name)


codes = PropertyDict((_codes[k], k) for k in _codes)


class HTTPError(IOError):
    """Something went wrong when making the request."""


class Request(urllib.request.Request):
    def __init__(self, method, url, **kwargs):
        self._method = method
        urllib.request.Request.__init__(self, url, **kwargs)

    def get_method(self):
        return self._method


class Response(object):
    def __init__(self, addinfourl, request):
        self._r = addinfourl
        self.request = request
        self.status_code = self._r.getcode()
        self.headers = self._r.headers
        self.cookies = self._read_cookies(self._r, request)
        self.url = self._r.geturl()
        self.content = self._r.read()

    @classmethod
    def _read_cookies(cls, response, request):
        jar = http_cookiejar.CookieJar()
        jar.extract_cookies(response, request)

        return {cookie.name: cookie.value for cookie in jar}

    @classmethod
    def _encoding_from_message(cls, message):
        if six.PY2:
            for value in message.getplist():
                if value[:8] == 'charset=':
                    return value[8:]
        else:
            return message.get_content_charset()

    def json(self, **kwargs):
        """Decodes response as JSON."""
        encoding = detect_encoding(self.content[:4])
        value = self.content.decode(encoding)

        return simplejson.loads(value, **kwargs)

    @property
    def text(self):
        encoding = self._encoding_from_message(self.headers)

        return self.content.decode(encoding)

    @property
    def links(self):
        """A dict of dicts parsed from the response 'Link' header (if set)."""
        # <https://example.com/?page=2>; rel="next", <https://example.com/?page=34>; rel="last"'
        # becomes
        # {
        #     'last': {'rel': 'last', 'url': 'https://example.com/?page=34'},
        #     'next': {'rel': 'next', 'url': 'https://example.com/?page=2'},
        # },
        result = {}
        if 'Link' in self.headers:
            value = self.headers['Link']

            for part in re.split(r', *<', value):
                link = {}
                vs = part.split(';')

                # First section is always an url.
                link['url'] = vs.pop(0).strip('\'" <>')

                for v in vs:
                    if '=' in v:
                        key, v = v.split('=')
                        link[key.strip('\'" ')] = v.strip('\'" ')

                rkey = link.get('rel') or link['url']
                result[rkey] = link

        return result

    @property
    def ok(self):
        try:
            self.raise_for_status()
        except HTTPError:
            return False
        else:
            return True

    def raise_for_status(self):
        """Raises HTTPError if the request got an error."""
        if 400 <= self.status_code < 600:
            message = 'Error %s for %s' % (self.status_code, self.url)
            raise HTTPError(message)


class HTTPErrorHandler(urllib.request.HTTPDefaultErrorHandler):
    def http_error_default(self, req, fp, code, msg, hdrs):
        # urllib2 raises an exception on 4xx and 5xx. Make us behave more like
        # requests.
        return fp


class HTTPRedirectHandler(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        # urllib2 follows redirects by default. This handler doesn't.
        return None


def detect_encoding(value):
    """Returns the character encoding for a JSON string."""
    # https://tools.ietf.org/html/rfc4627#section-3
    if six.PY2:
        null_pattern = tuple(bool(ord(char)) for char in value[:4])
    else:
        null_pattern = tuple(bool(char) for char in value[:4])

    encodings = {
        # Zero is a null-byte, 1 is anything else.
        (0, 0, 0, 1): 'utf-32-be',
        (0, 1, 0, 1): 'utf-16-be',
        (1, 0, 0, 0): 'utf-32-le',
        (1, 0, 1, 0): 'utf-16-le',
    }

    return encodings.get(null_pattern, 'utf-8')


def _build_opener(allow_redirects=True, verify=True):
    # We need a custom opener so we can choose to not follow redirects and
    # not treat 4xx and 5xx responses as errors.
    handlers = [HTTPErrorHandler]
    if not allow_redirects:
        handlers.append(HTTPRedirectHandler)

    if not verify:
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        handler = urllib.request.HTTPSHandler(context=ssl_context)
        handlers.append(handler)

    opener = urllib.request.build_opener(*handlers)

    return opener


def _encode_basic_auth(name, password):
    value = ('%s:%s' % (name, password)).encode(LATIN1).strip()
    value = b'Basic ' + base64.b64encode(value)

    return value


def _encode_data(data):
    """Handle a mapping or a sequence of pairs."""
    return urllib.parse.urlencode(data, doseq=True)


def _guess_filename(fileobj):
    name = getattr(fileobj, 'name', None)
    if name and name[:1] != '<' and name[-1:] != '>':
        return os.path.basename(name)


def _choose_boundary():
    chars = 'abcdefghijklmnopqrstuvwxyz123456789'
    boundary = ''.join(random.choice(chars) for _ in range(40))

    return boundary.encode('ascii')


def _build_form_data(data=None, files=None):
    # https://pymotw.com/2/urllib2/#uploading-files
    boundary = _choose_boundary()
    parts = []
    parts_boundary = b'--' + boundary

    # Has to be a dict if you are uploading files.
    if data:
        for field_name in sorted(data):
            parts.extend([
                parts_boundary,
                ('Content-Disposition: form-data; name="%s"' % field_name).encode(LATIN1),
                b'',
                data[field_name],
            ])

    if files:
        # files is a dict or list of 2-tuples, but the values can be
        # - a file-like object open for reading
        # - a pair of (filename, file-like object)
        # - a pair of (filename, byte string)
        if isinstance(files, dict):
            files = sorted(files.items())

        for field_name, name_and_file in files:
            if hasattr(name_and_file, 'read'):
                name = _guess_filename(name_and_file) or field_name
                value = name_and_file
            else:
                name, value = name_and_file

            content_type = (mimetypes.guess_type(name)[0] or BINARY_TYPE).encode(LATIN1)

            if not isinstance(value, six.binary_type):
                value = value.read()

            parts.extend([
                parts_boundary,
                ('Content-Disposition: file; name="%s"; filename="%s"' % (field_name, name)).encode(LATIN1),
                b'Content-Type: ' + content_type,
                b'',
                value,
            ])

    parts.extend([
        parts_boundary + b'--',
        b'',
    ])

    content_type = b'multipart/form-data; boundary=' + boundary
    data = b'\r\n'.join(parts)

    return content_type, data


def build_cookie(name, value):
    return http_cookiejar.Cookie(
        version=0,
        name=name,
        value=value,
        port=None,
        port_specified=False,
        domain='',
        domain_specified=False,
        domain_initial_dot=False,
        path='/',
        path_specified=False,
        secure=False,
        expires=None,
        discard=True,
        comment=None,
        comment_url=None,
        rest={'HttpOnly': None},
        rfc2109=False,
    )


def _merge_params(url, params):
    """Merge and encode query parameters with an URL."""
    if isinstance(params, dict):
        params = list(params.items())

    scheme, netloc, path, query, fragment = urllib.parse.urlsplit(url)
    url_params = urllib.parse.parse_qsl(query, keep_blank_values=True)
    url_params.extend(params)

    query = _encode_data(url_params)

    return urllib.parse.urlunsplit((scheme, netloc, path, query, fragment))


def _build_request(method, url, params=None, data=None, headers=None,
            cookies=None, auth=None, json=None, files=None):
    headers = {k.lower(): v for k, v in headers.items()} if headers else {}
    headers.setdefault('user-agent', _user_agent)

    if params:
        url = _merge_params(url, params)

    if auth:
        name, password = auth
        headers['authorization'] = _encode_basic_auth(name, password)

    if data and not files and not isinstance(data, six.binary_type):
        data = _encode_data(data)
        data = data.encode('ascii')

    if json:
        # If you send data and json, json overwrites data.
        data = simplejson.dumps(json).encode('utf-8')
        headers['content-type'] = JSON_TYPE

    if files:
        content_type, data = _build_form_data(data, files)
        headers['content-type'] = content_type

    request = Request(method, url, data=data, headers=headers)

    if cookies:
        jar = http_cookiejar.CookieJar()
        for key, value in cookies.items():
            cookie = build_cookie(key, value)
            jar.set_cookie(cookie)

        jar.add_cookie_header(request)

    return request


def _build_response(urllib_response, request):
    response = Response(urllib_response, request)

    return response


def request(method, url, params=None, data=None, headers=None, cookies=None,
            auth=None, json=None, files=None, allow_redirects=True, verify=True,
            timeout=None):
    request = _build_request(
        method,
        url,
        params=params,
        data=data,
        headers=headers,
        cookies=cookies,
        auth=auth,
        json=json,
        files=files,
    )

    _opener = _build_opener(allow_redirects=allow_redirects, verify=verify)

    # Better than trying to re-use urllib2's default timeout value. For regular
    # Python a timeout raises socket.timeout but App Engine will raise
    # google.appengine.api.urlfetch_errors.DeadlineExceededError.
    kwargs = {} if timeout is None else {'timeout': timeout}
    urllib_response = _opener.open(request, **kwargs)

    return _build_response(urllib_response, request)


delete = functools.partial(request, 'DELETE')
get = functools.partial(request, 'GET')
head = functools.partial(request, 'HEAD')
patch = functools.partial(request, 'PATCH')
post = functools.partial(request, 'POST')
put = functools.partial(request, 'PUT')
